### PR TITLE
Optimizations, part 2

### PR DESCRIPTION
**Pull request by hurricup**
Same issues in different places: multiple unnecessary instances of heavy objects, duplications. 

## Summary by CodeRabbit

- **New Features**
  - Enhanced equality and hash code functionalities for `ResourceBundleWithFallback`.
  - Introduced singleton pattern for `RuleFilterCreator` to improve resource management.
  - Added new methods in `MorfologikFrenchSpellerRule` for better rule instance handling.

- **Bug Fixes**
  - Streamlined resource bundle retrieval in various classes to improve performance and clarity.

- **Refactor**
  - Updated method implementations in the `French` class to simplify rule creation and management.
  - Improved import statements across multiple classes for better clarity.
  - Simplified instantiation logic in `MorfologikFrenchSpellerRuleTest`.
  - Enhanced efficiency in filter creation within `XMLRuleHandler`.

- **Tests**
  - Adjusted test initialization for `MorfologikFrenchSpellerRule` to use the new static method for rule retrieval.

### CODE REVIEW

File: languagetool-core/src/main/java/org/languagetool/ResourceBundleWithFallback.java [hunk 1]
```diff
@@ -49,4 +50,18 @@ public Enumeration<String> getKeys() {
     return bundle.getKeys();
   }
 
+  @Override
+  public final boolean equals(Object o) {
+    if (!(o instanceof ResourceBundleWithFallback)) return false;
+
+    ResourceBundleWithFallback that = (ResourceBundleWithFallback) o;
+    return Objects.equals(bundle, that.bundle) && Objects.equals(fallbackBundle, that.fallbackBundle);
+  }
```
[the remark this answers is not in the document]
Comment (SteVio89) [reply languagetool-core/src/main/java/org/languagetool/ResourceBundleWithFallback.java:59]: @hurricup please check this suggestion

File: languagetool-language-modules/fr/src/main/java/org/languagetool/language/French.java [hunk 2]
```diff
@@ -150,8 +150,8 @@ public Contributor[] getMaintainers() {
 
 
   @Override
-  public SpellingCheckRule createDefaultSpellingRule(ResourceBundle messages) throws IOException {
-    return new MorfologikFrenchSpellerRule(messages, this, null, Collections.emptyList());
+  public SpellingCheckRule createDefaultSpellingRule(ResourceBundle messages) {
+    return MorfologikFrenchSpellerRule.getRule(messages, this);
```
[the remark this answers is not in the document]
Comment (SteVio89) [reply languagetool-language-modules/fr/src/main/java/org/languagetool/language/French.java:154]: @hurricup please check this suggestion

**Review by SteVio89 — CHANGES_REQUESTED** [review]
Please see my comments

**Comment by hurricup** [comment]
Sure, didn't have a time yet

**Comment by SteVio89** [comment]
Thank you!
Didn't mean to rush you, just wasn't sure if you saw them.

File: languagetool-language-modules/fr/src/main/java/org/languagetool/language/French.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (hurricup) [reply languagetool-language-modules/fr/src/main/java/org/languagetool/language/French.java:154]: Fixed exception propagation.

**Review by SteVio89 — APPROVED** [review]

**Pull request merged**